src/Waargonaut/Generic.hs

```diff
@@ -513,6 +523,61 @@ gEncoder opts = Tagged . E.encodeA $ \a -> hcollapse $ hcliftA2
     pjE = Proxy :: Proxy (JsonEncode t)
     pt  = Proxy :: Proxy t
 
+-- | Create a 'Tagged' 'ObjEncoder' for type @ a @, tagged by @ t @.
+--
+-- This isn't compatible with the 'JsonEncode' typeclass because it creates an
+-- 'ObjEncoder' and for consistency reasons the 'JsonEncode' typeclass produces
+-- 'Encoder's.
+--
+-- However it lets you more easily access the 'Data.Functor.Contravariant.Contravariant'
+-- functionality that is part of the 'ObjEncoder' type.
+--
+-- @
+-- data Foo = Foo { fooA :: Text, fooB :: Int } deriving (Eq, Show)
+-- deriveGeneric ''Foo
+--
+-- objEncFoo :: Applicative f => ObjEncoder f Foo
+-- objEncFoo = untag $ gObjEncoder (defaultOps { _optionsFieldName = drop 3 })
+--
+-- @
+--
+gObjEncoder
+  :: forall t a f xs.
+     ( Generic a
+     , Applicative f
+     , HasDatatypeInfo a
+     , All2 (JsonEncode t) (Code a)
+     , IsRecord a xs
+     )
+  => Options
+  -> Tagged t (E.ObjEncoder f a)
+gObjEncoder opts = Tagged . E.objEncoder $ \a -> hcollapse $ hcliftA2
+  (Proxy :: Proxy (All (JsonEncode t)))
+  createObject
+  (jsonInfo (opts { _optionsNewtypeWithConsName = FieldNameAsKey }) (Proxy :: Proxy a))
+  (unSOP $ from a)
+  where
+    createObject :: ( All (JsonEncode t) ys
+                    , Applicative f
+                    )
+                 => JsonInfo ys
+                 -> NP I ys
+                 -> K (f (JObject WS Json)) ys
+    createObject (JsonRec _ fields) cs = K . pure .
+      foldl' (&) (_Empty # ()) . hcollapse $ hcliftA2 pjE toObj fields cs
+
+    createObject (JsonOne (Tag t)) (I a :* Nil) = K . pure $
+      E.onObj' (Text.pack t) (E.asJson' (T.proxy mkEncoder pt) a) E.json (_Empty # ())
+
+    -- IsRecord constraint should make this impossible.
+    createObject _ _ = error "impossible?"
```

Comment:
I suggest changing the message to something like.
`"The impossible has happened. Please report this as a waargonaut bug here: https://github.com/qfpl/waargonaut"`